subsys/net/lib/coap/coap_server.c

```diff
@@ -0,0 +1,749 @@
+		}
+	} else if (coap_remove_observer(resource, obs)) {
+		memset(obs, 0, sizeof(*obs));
+		return 1;
+	}
+
+	return 0;
+}
+
+static int coap_server_process(int sock_fd)
+{
+	uint8_t buf[CONFIG_COAP_SERVER_MESSAGE_SIZE];
+	struct sockaddr client_addr;
+	socklen_t client_addr_len = sizeof(client_addr);
+	struct coap_service *service = NULL;
+	struct coap_packet request;
+	struct coap_pending *pending;
+	struct coap_option options[MAX_OPTIONS] = { 0 };
+	uint8_t opt_num = MAX_OPTIONS;
+	uint8_t type;
+	ssize_t received;
+	int ret;
+
+	received = zsock_recvfrom(sock_fd, buf, sizeof(buf), ZSOCK_MSG_DONTWAIT, &client_addr,
+				  &client_addr_len);
+	__ASSERT_NO_MSG(received <= sizeof(buf));
+
+	if (received < 0) {
+		if (errno == EWOULDBLOCK) {
+			return 0;
+		}
+
+		LOG_ERR("Failed to process client request (%d)", -errno);
+		return -errno;
+	}
+
+	ret = coap_packet_parse(&request, buf, received, options, opt_num);
+	if (ret < 0) {
+		LOG_ERR("Failed To parse coap message (%d)", ret);
+		return ret;
+	}
+
+	(void)k_mutex_lock(&lock, K_FOREVER);
+	/* Find the active service */
+	COAP_SERVICE_FOREACH(svc) {
+		if (svc->data->sock_fd == sock_fd) {
+			service = svc;
+			break;
+		}
+	}
+	if (service == NULL) {
+		ret = -ENOENT;
+		goto unlock;
+	}
+
+	type = coap_header_get_type(&request);
+
+	pending = coap_pending_received(&request, service->data->pending, MAX_PENDINGS);
+	if (pending) {
+		uint8_t token[COAP_TOKEN_MAX_LEN];
+		uint8_t tkl;
+
+		switch (type) {
+		case COAP_TYPE_RESET:
+			tkl = coap_header_get_token(&request, token);
+			coap_service_remove_observer(service, NULL, &client_addr, token, tkl);
+			__fallthrough;
+		case COAP_TYPE_ACK:
+			coap_server_free(pending->data);
+			coap_pending_clear(pending);
+			break;
+		default:
+			LOG_WRN("Unexpected pending type %d", type);
+			ret = -EINVAL;
+			goto unlock;
+		}
+
+		goto unlock;
+	} else if (type == COAP_TYPE_ACK || type == COAP_TYPE_RESET) {
+		LOG_WRN("Unexpected type %d without pending packet", type);
+		ret = -EINVAL;
+		goto unlock;
+	}
+
+	if (IS_ENABLED(CONFIG_COAP_SERVER_WELL_KNOWN_CORE) &&
+	    coap_header_get_code(&request) == COAP_METHOD_GET &&
+	    coap_uri_path_match(COAP_WELL_KNOWN_CORE_PATH, options, opt_num)) {
+		uint8_t well_known_buf[CONFIG_COAP_SERVER_MESSAGE_SIZE];
+		struct coap_packet response;
+
+		ret = coap_well_known_core_get_len(service->res_begin,
+						   COAP_SERVICE_RESOURCE_COUNT(service),
+						   &request, &response,
+						   well_known_buf, sizeof(well_known_buf));
+		if (ret < 0) {
+			LOG_ERR("Failed to build well known core for %s (%d)", service->name, ret);
+			goto unlock;
+		}
+
+		ret = coap_service_send(service, &response, &client_addr, client_addr_len);
+	} else {
+		ret = coap_handle_request_len(&request, service->res_begin,
+					      COAP_SERVICE_RESOURCE_COUNT(service),
+					      options, opt_num, &client_addr, client_addr_len);
+
+		/* Shortcut for replying a code without a body */
+		if (ret > 0  && ret <= UINT8_MAX) {
```

Comment:
What's the meaning of `UINT8_MAX` in this case? Plus, I'm a bit confused regarding the documentation of `coap_handle_request_len()`, it only mentions that `0` is returned in case of success.

Reply:
Yes, this "feature" should be explained a bit more in the documentation.

Both `coap_handle_request` and `coap_handle_request_len` return the value of the handler called in the resource. I think it's very nice to have the option to send a response code (either success or error), without the need to build an acknowledge buffer in each handler. An example:

```c
static const struct gpio_dt_spec led = GPIO_DT_SPEC_GET_OR(DT_ALIAS(led0), gpios, {0});

static int led_put(struct coap_resource *resource, struct coap_packet *request,
		   struct sockaddr *addr, socklen_t addr_len)
{
	const uint8_t *payload;
	uint16_t payload_len;

	if (!gpio_is_ready_dt(&led)) {
		return COAP_RESPONSE_CODE_INTERNAL_ERROR;
	}

	payload = coap_packet_get_payload(request, &payload_len);
	if (payload_len != 1) {
		return COAP_RESPONSE_CODE_BAD_REQUEST;
	}

	if (gpio_pin_set_dt(&led, payload[0]) < 0) {
		return COAP_RESPONSE_CODE_INTERNAL_ERROR;
	}

	return COAP_RESPONSE_CODE_CHANGED;
}
```

With the shortcut feature, a response will be sent from the underlying server code.

Reply:
Should this be turned into a `Kconfig` option so users are aware? @rlubos

Reply:
I don't think there's a need for a Kconfig option here, personally I don't have an issue with this feature, as it seems reasonable, it'd just make sense to document it somewhere. Maybe we should elaborate on the expected method handlers behavior in `COAP_RESOURCE_DEFINE()` documentation? That if they don't need to include payload in the response, they can just return the response code to reply with.

I think it'd also make sense to extend the existing CoAP documentation we have (https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/connectivity/networking/api/coap.rst), or add a separate page (like for CoAP client helper lib), to raise awareness of this new CoAP service functionality (which btw seems pretty nice and helpful!).